larq_compute_engine/tflite/kernels/bconv2d.cc

```diff
@@ -342,81 +300,52 @@ float GetTensorValue(const TfLiteTensor* tensor, int index) {
-  for (int i = 0; i < params->channels_out; ++i) {
-    const double post_mul = GetTensorValue(post_activation_multiplier, i);
-    const double post_bias = GetTensorValue(post_activation_bias, i);
-
-    params->scaled_post_activation_multiplier[i] = post_mul / output_scale;
-    params->scaled_post_activation_bias[i] =
-        post_bias / output_scale + output_zero_point;
-  }
-}
-
 void OneTimeSetup(TfLiteContext* context, TfLiteNode* node,
                   TfLiteBConv2DParams* params) {
-  if (!params->is_padding_correction_cached &&
-      (params->padding_type == kTfLitePaddingSame && params->pad_value == 0)) {
-    const auto* filter = GetInput(context, node, 1);
-    const auto* post_activation_multiplier = GetInput(context, node, 2);
-
-    // Fill the zero-padding cache
-    ce::core::PaddingFunctor padding_functor;
+  const auto* filter = GetInput(context, node, 1);
+  const auto* post_activation_multiplier = GetInput(context, node, 2);
+  const auto* post_activation_bias = GetInput(context, node, 3);
+  const auto* output = GetOutput(context, node, 0);
 
-    std::size_t padding_cache_size = padding_functor.get_cache_size(
+  // For 'same-zero' padding, compute the padding-correction.
+  if (params->padding_type == kTfLitePaddingSame && params->pad_value == 0) {
+    params->padding_buffer.resize(ce::core::PaddingFunctor::get_cache_size(
         params->filter_height, params->filter_width, params->channels_out,
-        params->dilation_height_factor, params->dilation_width_factor);
-
-    params->padding_buffer.resize(padding_cache_size);
-
-    padding_functor.cache_correction_values(
+        params->dilation_height_factor, params->dilation_width_factor));
+    ce::core::PaddingFunctor::cache_correction_values(
         GetTensorData<TBitpacked>(filter), params->filter_height,
         params->filter_width, params->channels_out, params->channels_in,
         params->dilation_height_factor, params->dilation_width_factor,
         GetTensorData<float>(post_activation_multiplier),
         params->padding_buffer.data());
   }
-  params->is_padding_correction_cached = true;
-  if (!params->is_quantization_initialized) {
-    SetupQuantization(context, node, params);
-    params->is_quantization_initialized = true;
+
+  // For int8 output, compute the 'effective' multiplier/bias with fused output
+  // scale and zero-point.
+  if (output->type == kTfLiteInt8) {
+    params->scaled_post_activation_multiplier.resize(params->channels_out);
+    params->scaled_post_activation_bias.resize(params->channels_out);
+    const double output_zero_point =
+        static_cast<double>(output->params.zero_point);
+    const double output_scale = static_cast<double>(output->params.scale);
+    for (int i = 0; i < params->channels_out; ++i) {
+      const double post_mul = GetTensorValue(post_activation_multiplier, i);
+      const double post_bias = GetTensorValue(post_activation_bias, i);
```

Comment:
With the demand that the `post_activation_xxx` are always float, we can get rid of the `GetTensorValue` function and just write `GetTensorData<float>(post_activation_xxx)[i];` or something.